### PR TITLE
Refactor rename/migration handling in `Formulary`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Merge most of the code paths for loading with/without API. Instead of checking if the renamed/alias file exists, we simply assume so, as the end result is an error anyways if it doesn't.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me but would like to see ✅ from @Bo98 and/or @apainintheneck as they've been working here recently.

**Review by apainintheneck — APPROVED** [review]
This makes sense to me. From what I can tell `Formulary.loader_for` doesn't have [m]any tests but the loader code should be tested pretty well indirectly through all of the `Formulary.factory` tests so I'm not too worried about this change.

File: Library/Homebrew/formulary.rb [hunk 1]
```diff
@@ -902,47 +902,49 @@ def self.path(ref)
...
+    elsif (new_tap_name = tap.tap_migrations[name].presence)
+      new_tap_user, new_tap_repo, = new_tap_name.split("/")
+      new_tap_name = "#{new_tap_user}/#{new_tap_repo}"
+      new_tap = Tap.fetch new_tap_name
+      new_tap.ensure_installed!
+      new_tapped_name = "#{new_tap_name}/#{name}"
+      name, tap, = Formulary.tap_formula_name_type(new_tapped_name, warn: false)
+      old_name = tapped_name
+      new_name = new_tap.core_tap? ? name : new_tapped_name
+      type = :migration
```
Comment (apainintheneck) [inline Library/Homebrew/formulary.rb:927]: I guess the main difference in this logic at second glance is the removal of the `Formulary.find_formula_in_tap` and file path checks. Maybe those helped prevent infinite loops before and would have helped prevent what happened in https://github.com/Homebrew/brew/issues/16628.

CC: @dduugg

Comment (reitermarkus) [reply Library/Homebrew/formulary.rb:927]: Yes, previously it would just skip all this if the formula path exists.

Comment (dduugg) [reply Library/Homebrew/formulary.rb:927]: Thanks for the attention and fix here, @apainintheneck @reitermarkus 🥰

Comment (MikeMcQuaid) [reply Library/Homebrew/formulary.rb:927]: @reitermarkus @apainintheneck @dduugg Should we revert this PR given https://github.com/Homebrew/brew/issues/16628 is still open?

**Pull request merged**